portal-impl/src/com/liferay/portal/layoutconfiguration/util/RuntimePageImpl.java

```diff
@@ -446,10 +449,39 @@ protected StringBundler doProcessTemplate(
 			for (PortletRenderer portletRenderer : portletRenderers) {
 				Portlet portlet = portletRenderer.getPortlet();
 
+				if (!portlet.isReady()) {
+					continue;
+				}
+
+				InvokerPortlet invokerPortlet = null;
+
+				try {
+					invokerPortlet = PortletInstanceFactoryUtil.create(
+						portlet, request.getServletContext());
+				}
+				catch (PortletException pe) {
+					_log.error(pe, pe);
+				}
+				catch (RuntimeException re) {
+					_log.error(re, re);
+				}
+
+				if ((invokerPortlet == null) ||
+					!invokerPortlet.isHeaderPortlet()) {
```

Comment:
@ngriffin7a can we do the invokerPortlet.isHeaderPortlet() check without InvokerPortlet ? It should be doable by just using portlet, right? Since most portlet is not header portlet, all those InvokerPortlet is just a waste.

Reply:
@shuyangzhou I considered that but it would require executing a block of code similar to the following which has to first create a `PortletBag`:
https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portlet/internal/PortletInstanceFactoryImpl.java#L171-L196

So I thought that the solution in this PR was a little less messy. If you prefer that I do the `PortletBag` stuff in order to get the `rootInvokerPortletInstance` then please let me know and I will send you a follow-up PR.

Reply:
@ngriffin7a if the logic is significantly duplicated, pull it out as some kind of helper method, then use it in both locations. If it is just a bit similar or small enough, just let it be. Please resend this pull with that. Thanks!